File: 2164-two-best-non-overlapping-events/2164-two-best-non-overlapping-events.ts
function maxTwoEvents(events: number[][]): number {
    const eventsSortedByStartTime = [...events];
    eventsSortedByStartTime.sort(([s1], [s2]) => s2 - s1);
    const eventsSortedByEndTime = events;
    eventsSortedByEndTime.sort(([, e1], [, e2]) => e2 - e1);

    let maxRes = 0;
    let maxStart = 0;
    let startPtr = 0;

    for (let endPtr = 0; endPtr < eventsSortedByEndTime.length; ++endPtr) {
        while (startPtr < eventsSortedByStartTime.length && eventsSortedByStartTime[startPtr][0] > eventsSortedByEndTime[endPtr][1]) {
            maxStart = Math.max(maxStart, eventsSortedByStartTime[startPtr][2]);
            ++startPtr;
        }
        maxRes = Math.max(maxRes, maxStart + eventsSortedByEndTime[endPtr][2]);
    }
    return maxRes;
};

File: 2164-two-best-non-overlapping-events/2164-two-best-non-overlapping-events.rs
impl Solution {
    pub fn max_two_events(events: Vec<Vec<i32>>) -> i32 {
        let mut events_by_start_time = events.clone();
        events_by_start_time.sort_by(|e1, e2| e2[0].cmp(&e1[0]));
        let mut events_by_end_time = events.clone();
        events_by_end_time.sort_by(|e1, e2| e2[1].cmp(&e1[1]));

        let mut max_start = 0;
        let mut max_res = 0;
        let mut start_idx = 0;

        for e_end in events_by_end_time {
            while start_idx < events_by_start_time.len() && events_by_start_time[start_idx][0] > e_end[1] {
                max_start = std::cmp::max(events_by_start_time[start_idx][2], max_start);
                start_idx += 1;
            }
            max_res = std::cmp::max(e_end[2] + max_start, max_res);
        }
        max_res
    }
}